x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/token/TransportRefreshTokenAction.java

```diff
@@ -31,13 +30,12 @@ public TransportRefreshTokenAction(TransportService transportService, ActionFilt
 
     @Override
     protected void doExecute(Task task, CreateTokenRequest request, ActionListener<CreateTokenResponse> listener) {
-        tokenService.refreshToken(request.getRefreshToken(), ActionListener.wrap(tuple -> {
+        tokenService.refreshToken(request.getRefreshToken(), ActionListener.wrap(tokenResult -> {
             final String scope = getResponseScopeValue(request.getScope());
-            tokenService.authenticateToken(new SecureString(tuple.v1()), ActionListener.wrap(authentication -> {
-                listener.onResponse(new CreateTokenResponse(tuple.v1(), tokenService.getExpirationDelay(), scope, tuple.v2(), null,
-                    authentication));
-            },
-                listener::onFailure));
+            final CreateTokenResponse response =
+                new CreateTokenResponse(tokenResult.getAccessToken(), tokenService.getExpirationDelay(), scope,
+                    tokenResult.getRefreshToken(), null, tokenResult.getAuthentication());
```

Comment:
There are 3 authentication objects here:
1. The originating authentication
2. The authentication for which the token is created
3. The token's own authentication

We had brief discussion during the team meeting and agreed that above item 2 should be returned. However, we are returning item 3 here. The `tokenResult.getAuthentication()` is the authentication object of the newly created token, while No.2 is technically the old access token's authentication object.

I don't think it could cause any practical problem, in fact, No. 2 and 3 should be identical here. But it is strictly speaking an inconsistency. I am raising it here in case it's worth any discussion.

Reply:
The authentication type would be different, right? I also think we should pass `authentication` and not `tokenAuth` in the `CreateTokenResult` object

Reply:
I may have been unclear in my previous comment. Please let me clarify:

For other places, authencation 2 and 3 are indeed different, but not here. Also this is the only place where authentication 3 (`CreateTokenResult#getAuthentication`) is used. For all other places, authentication 2 is used, e.g. [TransportDelegatePkiAuthenticationAction](https://github.com/elastic/elasticsearch/pull/64031/files#diff-11f3471551a2ca238b9bbe5fbaed82727285d86b760cac14ec3bfbe312d2886bR87-R88), because authentication 2 is available without having to rely on `TokenService`.

In here, neither authentication 2 nor 3 is avilable, because we only have a refresh token. It needs to rely on `TokenService` to get both of them. `CreateTokenResult` only returns 3 and that is what gets used here. Luckily, 2 and 3 are no different in this case, so it is not an issue for now. But I wonder whether this equality would hold in future.

Reply:
> I also think we should pass authentication and not tokenAuth in the CreateTokenResult object

My comment was out of place too, I was reading `createOAuth2Tokens` when commenting.

> Luckily, 2 and 3 are no different in this case

As commented above too, The authentication type would be different, right? `innerRefresh` will call `createOAuth2Tokens` which will get the `Authentication` from the token document [and replace](https://github.com/elastic/elasticsearch/blob/e28dbde289ea7e7fbe6a81f89873d716ade33bd8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java#L324) the authentication type with `TOKEN` ?

Reply:
> As commented above too, The authentication type would be different, right? innerRefresh will call createOAuth2Tokens which will get the Authentication from the token document and replace the authentication type with TOKEN ?
>
The `Authentication` object parsed from a token document always has `TOKEN` as its authentication type. If I read the code correctly:
* `innerRefresh` [parses](https://github.com/elastic/elasticsearch/blob/e28dbde289ea7e7fbe6a81f89873d716ade33bd8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java#L1055-L1056) the token document to get its current `Authentication` object, which already has a `TOKEN` auth type.
* The above `Authentication` object is passed into `createOAuth2Tokens` and a new `Authentication` object is created from it by hardcoding a `TOKEN` auth type (as you linked above). But this is essentially a no-op because the original value is also `TOKEN`.

A token document is always created with [`TokenService#createTokenDocument`](https://github.com/elastic/elasticsearch/blob/e28dbde289ea7e7fbe6a81f89873d716ade33bd8/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java#L1476), which takes a `UserToken` and in turn its `Authentication` object. This `Authentication` object is guaranteed to have `TOKEN` auth type as you linked above.

Reply:
Thanks @ywangd for noticing! I do think having authentication for which the token's created is more logical, I missed it here (also, shouldn't make much difference for the perpose of this change). I have changed it use the authentication object representing the user for which the token was created